[ENH] adapting ROCKETClassifier to ROCKETRegressor

RavenRudi opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look
at our contribution guide: https://github.com/alan-turing-institute/sktime/blob/main/CONTRIBUTING.md
-->

#### Reference Issues/PRs
Contributes to #212



#### What does this implement/fix? Explain your changes.
Refactor ROCKETClassifier to ROCKETRegressor

#### Does your contribution introduce a new dependency? If yes, which one?
No.
<!--
If your contribution does add a new hard dependency, we may suggest to initially develop your contribution in a separate companion package in https://github.com/sktime/ to keep external dependencies of the core sktime package to a minimum.
-->

#### What should a reviewer concentrate their feedback on?

<!-- This section is particularly useful if you have a pull request that is still in development. You can guide the reviews to focus on the parts that are ready for their comments. We suggest using bullets (indicated by * or -) and filled checkboxes [x] here -->

#### Any other comments?
<!--
Please be aware that we are a loose team of volunteers so patience is necessary; assistance handling other issues is very welcome. We value all user contributions, no matter how minor they are. If we are slow to review, either the pull request needs some benchmarking, tinkering, convincing, etc. or more likely the reviewers are simply busy. In either case, we ask for your understanding during the review process.
-->

#### PR checklist
<!--
Please go through the checklist below. Please feel free to remove points if they are not applicable.
-->

##### For all contributions
- [ ] I've added myself to the [list of contributors](https://github.com/alan-turing-institute/sktime/blob/main/.all-contributorsrc).
- [ ] Optionally, I've updated sktime's [CODEOWNERS](https://github.com/alan-turing-institute/sktime/blob/main/CODEOWNERS) to receive notifications about future changes to these files.
- [ ] I've added unit tests and made sure they pass locally.

##### For new estimators
- [ ] I've added the estimator to the online documentation.
- [ ] I've updated the existing example notebooks or provided a new one to showcase how my estimator works.


<!--
Thanks for contributing!
-->

RavenRudi commented:
Hi @mloning,
I invested the day understanding the structure of the project and getting into OOP in python. It seems a bit different to the classical OOP languages such as Java but I think I'll get into it once I start coding more.
As for my refactoring task I got the impression that the core thing to do is changing `RidgeClassifierCV` to `RigdeCV`. Or did I miss something here? I have to admit that I did not read the rocket paper, so I am just scratching the surface. 
Next step would be refactoring all the duplicate code of `_rocket_regressor.py` and `_rocket_classifier.py` into a single class located somewhere in `sktime/series_as_features/estimators`, right?
I'd love to get some feedback if I'm heading in the right direction.

mloning commented:
Yes, that what I had in mind. If you want to learn more about ROCKET, this may be of interest: https://www.meetup.com/Berlin-Time-Series-Analysis-Meetup/events/276740583/

RavenRudi commented on sktime/series_as_features/base/estimators/shapelet_based/_rocket_estimator.py:
```diff
@@ -0,0 +1,163 @@
+        self.class_dictionary = {}
+
+        # We need to add is-fitted state when inheriting from scikit-learn
+        self._is_fitted = False
```
This line fixed it.

mloning replied:
Yes that shouldn't be necessary

mloning replied:
Does it work if we add a call to `super().__init__()` here

RavenRudi replied:
Yes, that fixed it!

RavenRudi commented:
Hi @mloning,

thank you for the nice chat yesterday! As discussed, I fixed the `RocketClassifier` by adding ` self._is_fitted = False` to `BaseROCKETEstimator` (commented that part in the review section).
However I do not understand why this fixed it... `_is_fitted` is also set to `False` in `BaseEstimator` which is a parent class of `ROCKETClassifier`. In my understanding, the `__init__()` method of `BaseEstimator` should be called somewhere in the MRO. So why is it necessary to set `_is_fitted` once again in `BaseROCKETEstimator`? 

On my local machine I get errors because there is no MrSEQLClassifier implemention, therefore the import fails. So I deleted that in order to run the unit tests properly. At least on my machine all tests pass now. Unfortunately the pipeline is not triggered here on GitHub.

Moreover I'm not quite sure if `_rocket_regressor` is tested, maybe you could check on that once again :)

Have a nice weekend!

mloning commented on sktime/regression/shapelet_based/_rocket_regressor.py:
```diff
@@ -0,0 +1,89 @@
+""" RandOm Convolutional KErnel Transform (ROCKET)
+"""
+
+__author__ = "RavenRudi"
```
```suggestion
__author__ = ["RavenRudi"]
```

mloning commented on sktime/regression/shapelet_based/_rocket_regressor.py:
```diff
@@ -0,0 +1,89 @@
+    ):
+
+        super().__init__(num_kernels, ensemble, ensemble_size, random_state)
+        super().__init__()
```
this line can be removed

mloning commented on sktime/series_as_features/base/estimators/shapelet_based/_rocket_estimator.py:
```diff
@@ -0,0 +1,163 @@
+"""
+
+__author__ = "Raven Rudi"
+# TODO
```
remove comment

mloning commented on sktime/series_as_features/base/estimators/shapelet_based/_rocket_estimator.py:
```diff
@@ -0,0 +1,163 @@
+        # We need to add is-fitted state when inheriting from scikit-learn
+        self._is_fitted = False
+
+    def fit(self, base_model, X, y):
```
rather than passing the `base_model` here as an argument I would define it as a class variable (see TSF refactoring for an example)

mloning replied:
`fit` is part of the common API, so we shouldn't change its signature

RavenRudi replied:
I solved that by making `BaseROCKETEstimator` an abstract class with a `base_estimator` as an abstract property. So the child classes are forced to provide a `base_estimator`. In the tsf refactoring, the parent class implicitly assumes that the child class provides a `base_estimator` . I don't think this is a neat structure :)

mloning reviewed (commented):
Thanks @RavenRudi - I left a few comments

RavenRudi commented:
Should be ready for review now @mloning. However, I'm still not sure if the `_rocket_regessor` is also tested.

mloning commented:
Sorry for the delay @RavenRudi - there is a related PR in #796, I'll come back to this one (hopefully won't take too long)

mloning commented:
Hi @RavenRudi, #796 is now merged, looks like you need to resolve some conflicts now, let me know if you need help with that. 

There may be some additional smaller changes implemented in #819, but that should be straightforward to fix.

RavenRudi commented:
Hi @mloning, I finally continued working on the issue. The test for `ROCKETClassifier` is passing, however I haven't written a test for `ROCKETRegressor`. If you could name me a dataset on which to test the regressor, I would give it a try.

mloning commented:
Hi @RavenRudi, great! We actually don't have a regression data set in the repo, so it would be good to add a small one. You may find some in this [paper](https://arxiv.org/abs/2006.12672). 

If made some final changes to #891, once that's merged, we can finish up this PR.

mloning commented:
Hi @RavenRudi there still seem to be some merge conflicts. Any luck with finding some regression data sets? This may be something to look into during the dev days in case you're attending!

RavenRudi commented:
Hi @mloning, I'm currently a bit under time pressure due to my thesis and an upcoming semester abroad, but I will be attending the dev days. Would be happy to have a look into that :). 
See you tomorrow!

RavenRudi commented:
> Hi @RavenRudi there still seem to be some merge conflicts. Any luck with finding some regression data sets? This may be something to look into during the dev days in case you're attending!

@mloning I just read the paper

> Hi @RavenRudi, great! We actually don't have a regression data set in the repo, so it would be good to add a small one. You may find some in this [paper](https://arxiv.org/abs/2006.12672).

@mloning Just read the paper. Would it make sense to implement a function similar to `load_UCR_UEA_dataset` for the [regression repository](http://tseregression.org/)? Since the datasets are available in .ts format, this should be doable. Let me know what you think. Would be happy to give it a try :)

mloning commented:
@RavenRudi yes that would be great! Adding one of them in sktime would also be a good idea. Check out the PR https://github.com/alan-turing-institute/sktime/pull/970 by @Riyabelle for instructions how to add data sets.

TonyBagnall commented:
that would be good, if you wanted to continue it would be good to have a function to load them directly from timeseriesclassification.com (it gives me an incentive to put them there)

RavenRudi commented:
Hi @TonyBagnall, @mloning,

just tried to add a regression dataset. I am not able to parse the provided .ts files. The `load_from_tsfile_to_dataframe` function throws an exception if the `@classLabel` tag is missing. I guess the function is only able to parse classification datasets.
I tried to use [this](https://github.com/ChangWeiTan/TS-Extrinsic-Regression/blob/master/utils/data_loader.py) modified function, but it didn't work either. 
Shall I just create a new PR that tries to extend the  `load_from_tsfile_to_dataframe` for regression datasets? Or am I missing something here?

mloning commented:
@RavenRudi sorry for the slow reply. 

Yes extending that function to handle regression data sets seems like a good idea to me, any thoughts @TonyBagnall? Does such a file format/loading function perhaps exist already somewhere? I found a version here: https://github.com/ChangWeiTan/TS-Extrinsic-Regression/blob/master/utils/data_loader.py but not sure if it does what we want.

mloning commented:
Hey @RavenRudi - how have you been? Shall we try to get this merged even without the regression data set? Would be a shame to lose the work in this PR.

RavenRudi commented:
> Hey @RavenRudi - how have you been? Shall we try to get this merged even without the regression data set? Would be a shame to lose the work in this PR.

Hi @mloning, I am good. Due to my starting exchange semester I didn't find the time to finish this off. I will have another look at it in two weeks - you will hear from me!

RavenRudi commented:
Hi @fkiraly,
here is my proposal for the design of the classification/regression module. It is influenced by its sklearn counterpart and sktime's forecaster interface.

`BaseTimeSeriesSupervisedLearner`  serves as a template for the base class of the respective algorithms, e.g ROCKET. The class tries to implement as much as possible at that point.
```python
# Template for XXXBaseEstimator
class BaseTimeSeriesSupervisedLearner(BaseEstimator, metaclass=ABC):
    
    def fit(self, X, y):

        coerce_to_numpy = self.get_tag("coerce-X-to-numpy", False)

        X, y = check_X_y(X, y, coerce_to_numpy=coerce_to_numpy)

        # Implementation of _fit in XXXBaseEstimator
        self._fit(X, y)

        # this should happen last
        self._is_fitted = True

        return self

    @abstractmethod
    def _fit(self,X,y):
        pass

    def predict(self, X):

        coerce_to_numpy = self.get_tag("coerce-X-to-numpy", False)

        X = check_X(X, coerce_to_numpy=coerce_to_numpy)
        self.check_is_fitted()

        # Implementation of _predict in XXXBaseEstimator
        y = self._predict(X)

        return y

    @abstractmethod
    def _predict(self, X):
        pass
```  

The `XXXBaseEstimator` class (e.g. ROCKET), then implements the missing `fit` and `predict` functionality.  Regression and classification might require slightly different implementations here and there. Hence it should be possible to identify whether the object is a regressor or a classifier, such that _fit and _predict can provide the correct functionality. 
```python
class XXXBaseEstimator(BaseTimeSeriesSupervisedLearner, metaclass=ABC):

    def _fit(self, X, y):
        # Concrete Implementation here
        # also somehow check if instantiated by regressor or classifier
    
    def _predict(self, X):
        # Concrete Implementation here, also check if regressor or classifier
```

The regressor and classifier classes are then exposed to the user.

```python
class XXXRegressor(RegressorMixin, XXXBaseEstimator):
    ## RegressorMixin is sklearn implementation and provides score function
    def __init__(*args):
        # only __init__ should be necessary in the most of the cases


class XXXClassifier(ClassifierMixin, XXXBaseEstimator):
    ## ClassifierMixin is sklearn implementation and provides score function
    def __init__(*args):
        # only __init__ should be necessary in the most of the cases

    def predict_proba(self, X):
```

fkiraly commented:
superseded by https://github.com/alan-turing-institute/sktime/pull/3126 which uses delegation.
